### PR TITLE
API reference

### LINKED ISSUE #46: Missing API function index in documentation & missing community guidelines
As requested by [JOSS](https://github.com/openjournals/joss-reviews/issues/7094), API function index and community guidelines are missing.

### PR DESCRIPTION
**Pull request by VeckoTheGecko**
This PR:
- adds an API page
- bugfixes link resolving function
- adds version dunder methods

Happy to discuss and edit layout as required. @michaeldenes could you review the generated docs and see it's as you expect? The docs are using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) but I'm not sure if all the docstrings adhere to that. I can install tooling to help lint that.

This closes #46

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -12,14 +12,17 @@
 import sys
 import warnings
 
+sys.path.insert(0, os.path.abspath('..')) # Allow to find `plasticparcels` package
```
Comment (VeckoTheGecko) [inline docs/conf.py:15]: This isn't necessary if a dev install is done of `plasticparcels`

**Comment by erikvansebille** [comment]
Hmm, getting an error when I try this code:
```bash
(parcels) erik ~/Codes/PlasticParcels/docs % make html  
Running Sphinx v7.3.7

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/erik/anaconda3/envs/parcels/lib/python3.12/site-packages/sphinx/config.py", line 509, in eval_config_file
    exec(code, namespace)  # NoQA: S102
    ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/erik/Codes/PlasticParcels/docs/conf.py", line 17, in <module>
    import plasticparcels
  File "/Users/erik/Codes/PlasticParcels/plasticparcels/__init__.py", line 5, in <module>
    from plasticparcels._version_setup import __version__, __version_tuple__  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'plasticparcels._version_setup'
```

I did just run
```
(parcels) erik ~/Codes/PlasticParcels % pip install --no-build-isolation --no-deps -e .
Obtaining file:///Users/erik/Codes/PlasticParcels
  Checking if build backend supports build_editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Building wheels for collected packages: plasticparcels
  Building editable for plasticparcels (pyproject.toml) ... done
  Created wheel for plasticparcels: filename=plasticparcels-0.0.0-0.editable-py3-none-any.whl size=5240 sha256=29281eedc8f7c6cbe06fe1372c1a59c4c95816fb6387a56911470499e3ae0af4
  Stored in directory: /private/var/folders/x0/1qxj679n55zcybylvdsb4bxh0000gq/T/pip-ephem-wheel-cache-hbca0z3n/wheels/46/7e/4c/6794e9b66cc3c539fe45e909a81eb509c8f932f180a62bb0a9
Successfully built plasticparcels
Installing collected packages: plasticparcels
Successfully installed plasticparcels-0.0.0
```
But that didn't help...

**Comment by VeckoTheGecko** [comment]
@erikvansebille I tried to recreate your error, but I couldn't. Perhaps its something to do with your conda environment? (e.g., having _parcels_ installed as a dev dependency?)

```
conda env create -n pp2 -f docs/environment_docs.yml
pip install --no-build-isolation --no-deps -e .
cd docs
make html
```

**Comment by erikvansebille** [comment]
Hmm, still doesn't work. See below

```bash
(parcels) erik ~/Codes/PlasticParcels % conda env create -n plasticparcels -f docs/environment_docs.yml
Channels:
 - conda-forge
 - defaults
Platform: osx-64
Collecting package metadata (repodata.json): done
Solving environment: done

Downloading and Extracting Packages:
                                                                                                                                                                           
Preparing transaction: done                                                                                                                                                
Verifying transaction: done                                                                                                                                                
Executing transaction: done                                                                                                                                                
#                                                                                                                                                                          
# To activate this environment, use                                                                                                                                        
#                                                                                                                                                                          
#     $ conda activate plasticparcels                                                                                                                                      
#                                                                                                                                                                          
# To deactivate an active environment, use                                                                                                                                 
#                                                                                                                                                                          
#     $ conda deactivate                                                                                                                                                   
                                                                                                                                                                           
(parcels) erik ~/Codes/PlasticParcels % conda activate plasticparcels
(plasticparcels) erik ~/Codes/PlasticParcels % pip install --no-build-isolation --no-deps -e .
Obtaining file:///Users/erik/Codes/PlasticParcels
  Checking if build backend supports build_editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Building wheels for collected packages: plasticparcels
  Building editable for plasticparcels (pyproject.toml) ... done
  Created wheel for plasticparcels: filename=plasticparcels-0.0.0-0.editable-py3-none-any.whl size=5238 sha256=3dfd01c2a1a18b0658739405d0d0bac82eb2780958d9bc91de5df9b62690a8fb
  Stored in directory: /private/var/folders/x0/1qxj679n55zcybylvdsb4bxh0000gq/T/pip-ephem-wheel-cache-yv031jol/wheels/46/7e/4c/6794e9b66cc3c539fe45e909a81eb509c8f932f180a62bb0a9
Successfully built plasticparcels
Installing collected packages: plasticparcels
Successfully installed plasticparcels-0.0.0
(plasticparcels) erik ~/Codes/PlasticParcels % cd docs
(plasticparcels) erik ~/Codes/PlasticParcels/docs % make html
Running Sphinx v5.3.0

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/erik/anaconda3/envs/plasticparcels/lib/python3.12/site-packages/sphinx/config.py", line 350, in eval_config_file
    exec(code, namespace)
  File "/Users/erik/Codes/PlasticParcels/docs/conf.py", line 17, in <module>
    import plasticparcels
  File "/Users/erik/Codes/PlasticParcels/plasticparcels/__init__.py", line 5, in <module>
    from plasticparcels._version_setup import __version__, __version_tuple__  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'plasticparcels._version_setup'

make: *** [html] Error 2
(plasticparcels) erik ~/Codes/PlasticParcels/docs % 
```

**Comment by VeckoTheGecko** [comment]
If you're encountering this error, its likely your install isn't behaving. Can you import `plasticparcels` if you open a Python prompt?

**Comment by VeckoTheGecko** [comment]
I have installed Ruff for this project. @michaeldenes this includes tooling to check your docstring format against [pydocstyle (D)](https://docs.astral.sh/ruff/rules/#pydocstyle-d) and [pydoclint (DOC)](https://docs.astral.sh/ruff/rules/#pydoclint-doc) rules. To use this tooling and see the errors you can do:

```
conda install pre_commit
pre-commit run --all-files
```

With the updated `pyproject.toml` config, your Ruff extension should pick up the errors and display them in VScode. If you instead want to disable the docstring checks do:

```diff
[tool.ruff.lint]
select = [
    "E",  # Error
    "F",  # pyflakes
    # "I",  # isort
    "B",  # Bugbear
    # "UP", # pyupgrade
    "LOG", # logging
    "ICN", # import conventions
    "G", # logging-format
    "RUF", # ruff
-   "D",  # pydocstyle
-   "DOC", # pydoclint
]
```
in `pyproject.toml`

**Comment by VeckoTheGecko** [comment]
I have also tried adding [pre-commit in CI](https://pre-commit.ci/). The owner of this repo should get a GitHub notification asking to authorise access.

**Comment by michaeldenes** [comment]
@VeckoTheGecko  this is looking good now. One nice to have would be a section navigation, like the `parcels` api reference, that allows the user to see 'constructors', 'kernels', etc.

**Pull request merged**